src/nanoarrow/nanoarrow_testing.hpp

```diff
@@ -1053,6 +1081,390 @@ class TestingJSONReader {
+            Check(value.contains("OFFSET"), error, "missing key 'OFFSET'"));
+        const auto& offset = value["OFFSET"];
+
+        if (array_view->layout.element_size_bits[buffer_i] == 32) {
+          NANOARROW_RETURN_NOT_OK(SetBufferInt<int32_t>(offset, buffer, error));
+        } else {
+          NANOARROW_RETURN_NOT_OK(SetBufferInt<int64_t>(offset, buffer, error));
+        }
+        break;
+      }
+
+      case NANOARROW_BUFFER_TYPE_DATA: {
+        NANOARROW_RETURN_NOT_OK(
+            Check(value.contains("DATA"), error, "missing key 'DATA'"));
+        const auto& data = value["DATA"];
+
+        switch (array_view->storage_type) {
+          case NANOARROW_TYPE_BOOL: {
+            nanoarrow::UniqueBitmap bitmap;
+            NANOARROW_RETURN_NOT_OK(SetBufferBitmap(data, bitmap.get(), error));
+            ArrowBufferMove(&bitmap->buffer, buffer);
+            return NANOARROW_OK;
+          }
+          case NANOARROW_TYPE_INT8:
+            return SetBufferInt<int8_t>(data, buffer, error);
+          case NANOARROW_TYPE_UINT8:
+            return SetBufferInt<uint8_t>(data, buffer, error);
+          case NANOARROW_TYPE_INT16:
+            return SetBufferInt<int16_t>(data, buffer, error);
+          case NANOARROW_TYPE_UINT16:
+            return SetBufferInt<uint16_t>(data, buffer, error);
+          case NANOARROW_TYPE_INT32:
+            return SetBufferInt<int32_t>(data, buffer, error);
+          case NANOARROW_TYPE_UINT32:
+            return SetBufferInt<uint32_t>(data, buffer, error);
+          case NANOARROW_TYPE_INT64:
+            return SetBufferInt<int64_t>(data, buffer, error);
+          case NANOARROW_TYPE_UINT64:
+            return SetBufferInt<uint64_t, uint64_t>(data, buffer, error);
+
+          case NANOARROW_TYPE_FLOAT:
+            return SetBufferFloatingPoint<float>(data, buffer, error);
+          case NANOARROW_TYPE_DOUBLE:
+            return SetBufferFloatingPoint<double>(data, buffer, error);
+
+          case NANOARROW_TYPE_STRING:
+            return SetBuffersString<int32_t>(data, ArrowArrayBuffer(array, buffer_i - 1),
+                                             buffer, error);
+          case NANOARROW_TYPE_LARGE_STRING:
+            return SetBuffersString<int64_t>(data, ArrowArrayBuffer(array, buffer_i - 1),
+                                             buffer, error);
+          case NANOARROW_TYPE_BINARY:
+            return SetBuffersBinary<int32_t>(data, ArrowArrayBuffer(array, buffer_i - 1),
+                                             buffer, error);
+          case NANOARROW_TYPE_LARGE_BINARY:
+            return SetBuffersBinary<int64_t>(data, ArrowArrayBuffer(array, buffer_i - 1),
+                                             buffer, error);
+          case NANOARROW_TYPE_FIXED_SIZE_BINARY:
+            return SetBuffersBinary<int64_t>(
+                data, nullptr, buffer, error,
+                array_view->layout.element_size_bits[buffer_i] / 8);
```

Comment:
It would probably be more readable if there was a separate helper function for fixed-size binary?